test/fixtures/linter/autofix/ansi/028_l003_lint_unreached_code/after.sql

```diff
@@ -0,0 +1,7 @@
+{% if True %}
```

Comment:
@alanmcruickshank: Could you take a look at the failing test and let me know what you think?

The linter returns a proper list of the issues in both parts of the file (reachable and unreachable):
```
$ sqlfluff lint test/fixtures/linter/autofix/ansi/028_l003_lint_unreached_code/before.sql
== [test/fixtures/linter/autofix/ansi/028_l003_lint_unreached_code/before.sql] FAIL                                                                                                                 
L:   2 | P:   1 | L003 | Expected indent of 4 spaces.
L:   5 | P:   1 | L003 | Expected indent of 4 spaces.
L:   6 | P:   1 | L003 | Expected indent of 4 spaces.
All Finished 📜 🎉!
```

But when running `sqlfluff fix`, only the first fix is applied, resulting in:
```
{% if True %}
    SELECT 1
{% else %}
    SELECT c
FROM t
WHERE c < 0
{% endif %}
```

My intuition is that the differing `sliced_file` contents and/or related things like placeholder segments prevent applying all the fixes directly. _If_ that's the case, one possible workaround would be to apply the fixes separately and combine the results.

I'd like to get your thoughts on this before beginning a deeper dive.

Reply:
Linking this up to what I posted on slack:

> Without actually running the code I'm not 100% sure, but I have a theory on why this might happen. When running the fix logic - the fixes don't patch the file - they overwrite it with a reconstructed version. So if we read the file twice, fix it twice and then try and write them again - only the fixes from one branch will be applied. The others with just be overwritten by the second branch which is written.

>
 In my mind, there are three ways around this problem.

> 1. Change the fix logic to try and patch rather than overwrite. Deep down - I think this is a losing battle, because all the references will be wrong if the file has been also modified by the other branch.
> 2. If we're in a fixing mode - explicitly lint the other branch after it has already been fixed by the first. This is probably the easier to implement - but feels clunky given the UX.
> 3. It might be possible to somehow condense or coalesce the two results into one file before writing. This is probably my preferred option. Details...

> If neither branch has applied their fixes - then the two `TemplatedFile` references should be compatible. They won't be identical (because you'll have fiddled the source file to force the branch evaluation) - but you'd be able to check whether they were otherwise identical. With a little finesse, you could take two `LintedFile` references, both with a set of fixes, different fixed files, but compatible raw files, and turn them into one entity with a combined set of fixes, which would enable the reconstructed file to contain both. You're effectively making a combined result file - before doing any interaction with the filesystem - and then only writing once.

> Logic would be something like:
> 1. Lint, fix etc... get as far as where we try to apply fixes.
> 2. In the fix application logic, (or perhaps just before), we group together any `LintedFile` objects which reference the same source file.
> 3. Identify which is the unaltered variant, and use that as a the base model.
> 4. Identify sections of the raw unaltered variant which aren't covered in the rendered version (code which you've already written - so maybe this needs to be encoded into a structure which can be used again later) - and importantly we also need to be able to map the location of these sections in both the raw source file and the altered source file.
Somewhere in the file reconstruction logic - whenever we reach an unrendered section in the unaltered variant - we look in the other variants to see if they want to make any changes there (once changes are remapped) and then bring them into the reconstructed source file at that stage.
> 5. File the fixed file once using this combined version.

> the place you'd need to make most of these changes would probably be in `LintedFile.fix_string` - although this logic would now need to be "multi-variant-aware". We still need to run `_generate_source_patches()` for each variant.
> The edits would need to then make `_slice_source_file_using_patches` and `_build_up_fixed_source_string` work with patches from multiple variants to then construct the combined file.